src/main/java/org/threadly/util/debug/Profiler.java

```diff
@@ -188,7 +194,81 @@ public void start() {
+   */
+  public ListenableFuture<String> start(long sampleDurationInMillis) {
+    return start(null, sampleDurationInMillis);
+  }
+  
+  /**
+   * Starts the profiler running in a new thread.  
+   * 
+   * If this profiler had previously ran, and is now sitting in a stopped state again.  The 
+   * statistics from the previous run will still be included in this run.  If you wish to clear 
+   * out previous runs you must call {@link #reset()} first.  
+   * 
+   * If an executor is provided, this call will block until the the profiler has been started on 
+   * the provided executor.
+   * 
+   * If {@code sampleDurationInMillis} is greater than zero the Profiler will invoke 
+   * {@link #stop()} in that many milliseconds.  If a provided executor is also a 
+   * {@link SimpleSchedulerInterface} then this will be used to schedule out when the profiler 
+   * should be stopped.
+   * 
+   * The returned {@link ListenableFuture} will be provided the dump when {@link #stop()} is 
+   * invoked next.  Either from a timeout provided to this call, or a manual invocation of 
+   * {@link #stop()}.
+   * 
+   * @param executor executor to execute on, or {@code null} if new thread should be created
+   * @param sampleDurationInMillis if greater than {@code 0} the profiler will only run for this many milliseconds
+   * @return Future that will be completed with the dump string when the profiler is stopped
+   */
+  public ListenableFuture<String> start(Executor executor, long sampleDurationInMillis) {
+    SettableListenableFuture<String> result = new SettableListenableFuture<String>();
+    start(executor, sampleDurationInMillis, result);
+    return result;
+  }
+  
+  /**
+   * The ultimate start call for the profiler.  This handles all possible start permutations.  
+   * If an {@link Executor} is provided, that will be used to run the profiler thread.  If there 
+   * is a duration provided > 0, a thread will be started to perform profiler shutdown.  If a 
+   * provided executor is also a {@link SimpleSchedulerInterface} then that will be used for 
+   * scheduling a task to stop the profiler.  If {@code sampleDurationInMillis} is greater than 
+   * {@code 0} and the profiler is already running, it will be stopped and restarted.
+   * 
+   * @param executor Executor or scheduler to use for execution as possible
+   * @param sampleDurationInMillis if greater than {@code 0} an automatic stop will occur after that many milliseconds 
+   * @param completionFuture If not {@code null} future will be called once the next {@link #stop()} is invoked
+   */
+  private void start(Executor executor, final long sampleDurationInMillis, 
+                     SettableListenableFuture<String> completionFuture) {
     synchronized (startStopLock) {
+      if (sampleDurationInMillis > 0) {
+        // stop in case it's running, this allows us to simplify our start logic
+        stop();
```

Comment:
I would argue that we might want to throw here.  If someone starts 2 times, i think they are just doing something bad in code and we should just not allow it.  So we will complete the first one instead of the last one.